File: py/packages/genkit-ai/src/genkit/blocks/tools.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

from typing import Any

from genkit.core.action import ActionRunContext
from genkit.typing import Part, ToolRequestPart, ToolResponse


class ToolRunContext(ActionRunContext):
    """Context for an tool execution."""

    def __init__(
        self,
        ctx: ActionRunContext,
    ):
        super().__init__(
            on_chunk=ctx._on_chunk if ctx.is_streaming else None,
            context=ctx.context,
        )

    def interrupt(self, metadata: dict[str, Any] | None = None):
        """Interrupts current execution of the tool."""
        raise ToolInterruptError(metadata=metadata)


# TODO: make this extend GenkitError once it has INTERRUPTED status
class ToolInterruptError(Exception):
    """Error throw to interrupt tool execution. Used for tool flow control."""

    def __init__(self, metadata: dict[str, Any]):
        self.metadata = metadata


def tool_response(
    interrupt: Part | ToolRequestPart,
    responseData: Any | None = None,
    metadata: dict[str, Any] | None = None,
) -> Part:
    """Constructs a tool response for an interrupted request."""
    # TODO: validate against tool schema
    tool_request = interrupt.root.tool_request if isinstance(interrupt, Part) else interrupt.tool_request
    return Part(
        tool_response=ToolResponse(
            name=tool_request.name,
            ref=tool_request.ref,
            output=responseData,
        ),
        metadata={
            'interruptResponse': metadata if metadata else True,
        },
    )

File: js/testapps/format-tester/src/tools.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { googleAI } from '@genkit-ai/google-genai';
import { genkit, z } from 'genkit';

const ai = genkit({
  plugins: [googleAI()],
  model: googleAI.model('gemini-2.5-flash'),
});

const lookupUsers = ai.defineTool(
  {
    name: 'lookupUsers',
    description: 'use this tool to list users',
    outputSchema: z.array(z.object({ name: z.string(), id: z.number() })),
  },
  async () => [
    { id: 123, name: 'Michael Bleigh' },
    { id: 456, name: 'Pavel Jbanov' },
    { id: 789, name: 'Chris Gill' },
    { id: 1122, name: 'Marissa Christy' },
  ]
);

async function main() {
  const { stream } = await ai.generateStream({
    prompt:
      'use the lookupUsers tool and generate silly nicknames for each, then generate 50 fake users in the same format. return a JSON array.',
    output: {
      format: 'json',
      schema: z.array(
        z.object({ id: z.number(), name: z.string(), nickname: z.string() })
      ),
    },
    tools: [lookupUsers],
  });

  for await (const chunk of stream) {
    console.log('raw:', chunk);
    console.log('output:', chunk.output);
  }
}
main();
